Add tax domain

arjun289 opened the pull request:
## Why?
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- If it delivers a story on the Pivotal tracker, please link to it here. -->
At present tax needs to be handled for product and shipping. The PR adds tax domain module which handles all the business logic and edge cases related to tax and is used in respective domain modules for package item and package to set taxes for product and shipping. 

## This change addresses the need by:
<!--- List and detail all changes made in this PR. -->
- Adds tax_class field to products so now all the products have a belongs to relationship with tax_class.
```
product bleongs_to tax_class
```
This relationship allows tax calculation for a product by findind respective values in a tax rate for that particular tax_class.
- Unit price field has been added to `package_item` this was done so as to keep the actual price after tax calculation which is exclusive of the tax.
- Adds tax domain to handle all the business logic. At present tax domain handles tax calculation for 
product and shipping.

[delivers #163283496]

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read [CONTRIBUTING.md][contributing].
- [x] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [x] I have updated the documentation wherever necessary.
- [x] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

arjun289 committed:
Make changes to product and package item for tax.

To allow products to be taxable, they should belong to a tax class.
Also, to handle order amount tally it was required to keep unit price
in package items which was exclusive of tax. Run time calculation could
have been used to calculate taxes but, it will lead to discrepancies
if tax config is modified.

This change
- Associated tax class with product.
- Also, tax class should be associated only with parent product and not
 variants, so changes were made to cater for that in product schema,
 controller and form template.
- Added unit price to package item.

arjun289 committed:
Add Domain For Tax Calculation

Tax calculations are performed based on multiple factors like,
tax zone, tax rates in the zone and address set in tax config.
Also, while calculating tax it has to be kept in mind
whether the prices entered are exclusive or inclusive of the
taxes.

This change
- Adds tax domain to handle tax calculation for package item and
  shipping cost.
- Modified `shipment.ex` to calculate taxes and unit prices for
  package item.
- Modified package domain to handle adding shipping cost and taxes.
- All the failing tests were modified to cater for tax related
  changes.

arjun289 committed:
Refactor Tax Domain

Refactored tax domain to add checks:
- If tax zone is not found with state then country
  should be tried.
- Also, remove check for shipping id not set as shipping
 id is required param for tax config so it was not necessary
- More tests were added to cover edge cases.

pkrawat1 commented on apps/snitch_core/lib/core/data/model/product.ex:
```diff
@@ -549,4 +549,35 @@ defmodule Snitch.Data.Model.Product do
+  In case supplied product is not a variant, an error tuple is
+  returned.
+  """
+  @spec get_parent_product(Product.t()) :: Product.t() | nil
```
`doc` and function return are different

pkrawat1 commented on apps/snitch_core/lib/core/domain/tax.ex:
```diff
@@ -0,0 +1,189 @@
+  def calculate(:package_item, line_item, order, stock_location) do
+    line_item = line_item |> Repo.preload(:product)
+
+    tax_class_id = line_item.product |> Product.get_tax_class_id()
```
avoid single pipe. Instead parameter should be used.

pkrawat1 commented on apps/snitch_core/priv/repo/migrations/20190131112645_modify_prod_pack_item_for_tax.exs:
```diff
@@ -0,0 +1,13 @@
+    end
+
+    alter table("snitch_package_items") do
+      add(:unit_price, String.to_atom("money_with_currency"))
```
string to atom ?

arjun289 replied:
Actually I am trying to follow the convention that has been set for all money types in migrations. 
It was added in this [PR 302](https://github.com/aviacommerce/avia/pull/302/files)

pkrawat1 reviewed (commented):
DID basic syntax checking only

arjun289 committed:
PR review fixes

- Removed single pipe chains
- Sorted out errors in docs.

arjun289 commented on apps/snitch_core/lib/core/data/schema/package_item.ex:
```diff
@@ -59,10 +59,18 @@ defmodule Snitch.Data.Schema.PackageItem do
+  The actual unit price for the package item. The `unit_price` may be same
+  or different from the `line_item` unit price depending on whether, line
+  item price is inclusive or exclusive of tax. In case line item price
+  is inclusive, this price is set after removing the tax amount form that.
```
typo form instead of that write line_item unit price

arjun289 commented on apps/snitch_core/lib/core/domain/package_item.ex:
```diff
@@ -15,4 +16,12 @@ defmodule Snitch.Domain.PackageItem do
+
+  @doc """
+  Returns the unit price and tax for a package item in the following
+  format %{unit_price: data, tax: data}
```
check whether correct?

arjun289 commented on apps/snitch_core/test/domain/order/transitions_test.exs:
```diff
@@ -161,13 +171,17 @@ defmodule Snitch.Domain.Order.TransitionsTest do
     setup :zones
     setup :shipping_methods
     setup :embedded_shipping_methods
+    setup :states
```
check if it can be pulled out.

arjun289 committed:
PR fixes

- Removed typo and errors in doc.
- Moved repetitive tasks for order setup in
  transition to a function.

ashish173 committed:
Merge branch 'develop' into tax-domain
